invenio_stats/aggregations.py

```diff
@@ -384,27 +422,35 @@ def run(self, start_date=None, end_date=None, update_bookmark=True):
         # If no events have been indexed there is nothing to aggregate
         if not dsl.Index(self.event_index, using=self.client).exists():
             return
-
+        if update_bookmark:
+            previous_bookmark = self.bookmark_api.get_bookmark()
+        else:
+            previous_bookmark = None
         lower_limit = (
-            start_date
-            or self.bookmark_api.get_bookmark()
-            or self._get_oldest_event_timestamp()
+            start_date or previous_bookmark or self._get_oldest_event_timestamp()
         )
         # Stop here if no bookmark could be estimated.
         if lower_limit is None:
             return
 
-        upper_limit = self._upper_limit(end_date, lower_limit)
+        upper_limit = self._upper_limit(end_date)
         dates = self._split_date_range(lower_limit, upper_limit)
+        # Let's get the timestamp before we start the aggregation.
+        # This will be used for the next iteration. Some events might be processed twice
+        self.bookmark_api.prepare_bookmark(end_date)
```

Comment:
It looks to me we can avoid the `new_timestamp` class variable, which is actually a state.
```suggestion
        end_date = end_date or (datetime.utcnow() - timedelta(minutes=1)).strftime(
                "%Y-%m-%dT%H:%M:%S"
            )
```
Then we can pass `end_date` to `commit_bookmark()`. Or something similar.

Reply:
I had something like this at the beginning. The part that I didn't like is that the magic of removing the timedelta had to be implemented them by whatever uses the bookmark. I thought it would be better to hide that in the bookmark class.  I meant to do it an object variable, not a class variable. Did I mess that up?

Reply:
Ok, `self.timestamp` object removed. The logic of substracting the timedelta has been moved to the get_bookmark